src/Data/Singletons/Prelude/Show.hs

```diff
@@ -0,0 +1,229 @@
+
+  showSpace :: SymbolS
+  showSpace = \xs -> " " <> xs
+
+  showCommaSpace :: SymbolS
+  showCommaSpace = showString ", "
+
+  appPrec, appPrec1 :: Nat
+  appPrec  = 10
+  appPrec1 = 11
+
+  instance Show a => Show [a] where
+    showsPrec _ = showList
+
+  -- -| This is not an ideal Show instance for Symbol, since the Show instance
+  -- for String escapes special characters. Unfortunately, GHC lacks the ability
+  -- to case on individual characters in a Symbol (at least, not without GHC
+  -- plugins), so this is the best we can do for the time being.
+  instance Show Symbol where
+    showsPrec _ = showString
+
+  show_tuple :: [SymbolS] -> SymbolS
+  show_tuple ss = showChar "("
+                . foldr1 (\s r -> s . showChar "," . r) ss
+                . showChar ")"
+
+  instance (Show a, Show b) => Show (a,b)  where
+    showsPrec _ (a,b) s = show_tuple [shows a, shows b] s
+
+  instance (Show a, Show b, Show c) => Show (a, b, c) where
+    showsPrec _ (a,b,c) s = show_tuple [shows a, shows b, shows c] s
+
+  instance (Show a, Show b, Show c, Show d) => Show (a, b, c, d) where
+    showsPrec _ (a,b,c,d) s = show_tuple [shows a, shows b, shows c, shows d] s
+
+  instance (Show a, Show b, Show c, Show d, Show e) => Show (a, b, c, d, e) where
+    showsPrec _ (a,b,c,d,e) s = show_tuple [shows a, shows b, shows c, shows d, shows e] s
+
+  instance (Show a, Show b, Show c, Show d, Show e, Show f) => Show (a,b,c,d,e,f) where
+    showsPrec _ (a,b,c,d,e,f) s = show_tuple [shows a, shows b, shows c, shows d, shows e, shows f] s
+
+  instance (Show a, Show b, Show c, Show d, Show e, Show f, Show g)
+          => Show (a,b,c,d,e,f,g) where
+    showsPrec _ (a,b,c,d,e,f,g) s
+          = show_tuple [shows a, shows b, shows c, shows d, shows e, shows f, shows g] s
+  |])
+
+-- At the moment, I can't think of a way to implement a PShow Nat instance,
+-- since that would require a way to convert Nats to Symbols at the type level.
+-- Oh well.
+
+instance SShow Nat where
+  sShowsPrec _ sn sx =
+    let n = fromSing sn
+        x = fromSing sx
+        ex = someSymbolVal (P.show n ++ T.unpack x)
+    in
+    case ex of
+      SomeSymbol (_ :: Proxy s) -> unsafeCoerce (SSym :: Sing s)
+
+  -- Annoyingly enough, the default definitions for sShow' and sShowList won't
+  -- typecheck because they rely on the type-level Show' and ShowList reducing,
+  -- but since we don't have a PShow Nat instance, that doesn't happen. To work
+  -- around this, we define sShow' and sShowList by hand.
+
+  sShow' sn =
+    let n = fromSing sn
+        ex = someSymbolVal (P.show n)
+    in
+    case ex of
+      SomeSymbol (_ :: Proxy s) -> unsafeCoerce (SSym :: Sing s)
+
+  sShowList sl sx =
+    let l = fromSing sl
+        x = fromSing sx
+        ex = someSymbolVal (P.show l ++ T.unpack x)
+    in
+    case ex of
+      SomeSymbol (_ :: Proxy s) -> unsafeCoerce (SSym :: Sing s)
+
+$(singShowInstances [ ''(), ''Maybe, ''Either, ''NonEmpty, ''Bool, ''Ordering ])
```

Comment:
I manually defined a `Show [a]` instance above. I didn't derive it since lists get shown specially.